File: kt/main/kotlin/moduleconfig/pushtofunction/fs/boot/FooSflBoot.kt
/*
 * Copyright © 2022 Paulo Villela. All rights reserved.
 * Use of this source code is governed by the MIT license
 * that can be found in the LICENSE file.
 */

package tryout.moduleconfig.pushtofunction.fs.boot

import tryout.moduleconfig.pushtofunction.fs.FooSflCfgInfo
import tryout.moduleconfig.pushtofunction.fs.FooSflCfgSrc
import tryout.moduleconfig.pushtofunction.fs.FooSflT
import tryout.moduleconfig.pushtofunction.fs.fooSflC
import tryout.moduleconfig.pushtofunction.config.AppCfgInfo
import tryout.moduleconfig.pushtovar.fwk.liftToNullary

private fun fooSflCfgAdapter0(appCfg: AppCfgInfo): FooSflCfgInfo {
	return FooSflCfgInfo(
		x = appCfg.x,
	)
}

var fooSflCfgAdapter = liftToNullary(::fooSflCfgAdapter0)

fun fooSflBoot(appCfg: (() -> AppCfgInfo)?): FooSflT {
	val fooSflCfgSrc = FooSflCfgSrc(
		get = fooSflCfgAdapter(appCfg),
		bar = barBfBoot(appCfg)
	)
	return fooSflC(fooSflCfgSrc)
}

File: rs/pushtofunction_old/src/fs/boot/foo_sfl_boot.rs
use crate::config::AppCfgInfo;
use crate::fs::{foo_sfl_c, FooSflCfgInfo, FooSflCfgSrc, FooSflT};
use crate::fwk::const_or_adapt_by_ref;
use std::sync::{Arc, OnceLock};

use super::bar_bf_boot;

fn foo_sfl_cfg_adapter(app_cfg: &AppCfgInfo) -> FooSflCfgInfo {
    FooSflCfgInfo {
        a: app_cfg.x.clone(),
        b: app_cfg.y,
    }
}

pub static FOO_SFL_CFG_INFO_OVERRIDE: OnceLock<FooSflCfgInfo> = OnceLock::new();

pub fn foo_sfl_boot(app_cfg: fn() -> Arc<AppCfgInfo>) -> FooSflT {
    let get = const_or_adapt_by_ref(
        FOO_SFL_CFG_INFO_OVERRIDE.get(),
        app_cfg,
        foo_sfl_cfg_adapter,
    );
    let foo_sfl_cfg_src = FooSflCfgSrc {
        get,
        bar: bar_bf_boot(app_cfg),
    };
    foo_sfl_c(foo_sfl_cfg_src)
}
